dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -311,15 +311,23 @@ ExchangeReceiverBase<RPCContext>::ExchangeReceiverBase(
 template <typename RPCContext>
 ExchangeReceiverBase<RPCContext>::~ExchangeReceiverBase()
 {
-    setState(ExchangeReceiverState::CLOSED);
-    msg_channel.finish();
+    close();
     thread_manager->wait();
 }
 
 template <typename RPCContext>
 void ExchangeReceiverBase<RPCContext>::cancel()
 {
-    setState(ExchangeReceiverState::CANCELED);
+    if (!setEndState(ExchangeReceiverState::CANCELED))
+        return;
```

Comment:
```suggestion
    setEndState(ExchangeReceiverState::CANCELED);
```

it's ok for finishing a queue twice.